File: src/extForm.jl
# This is the extensive formulation of PERT (1st case)
# the disruption does not affect the activities that have not been started

function obtainExtSols(mext,ω)
    solve(mext);
    texts = Dict();
    xexts = Dict();
    gexts = Dict();
    sexts = Dict();
    for i in pData.II
        texts[i] = getvalue(mext[:t][i,ω]);
        gexts[i] = getvalue(mext[:G][i,ω]);
        for j in pData.Ji[i]
            xexts[i,j] = getvalue(mext[:x][i,j,ω]);
            sexts[i,j] = getvalue(mext[:s][i,j,ω]);
        end
    end
    return texts,xexts,gexts,sexts;
end

function extForm(pData,disData,Ω,TL = Inf)
    M = Dict();
    for ω in Ω
        #M[ω] = sum(max(pData.D[i],pData.D[i]+disData[ω].d[i]) for i in pData.II if i != 0);
        M[ω] = 300;
    end

    mp = Model(solver = GurobiSolver(IntFeasTol = 1e-7, TimeLimit = TL));
    # mp = Model(solver = CplexSolver(CPX_PARAM_EPRHS = 1e-7,CPX_PARAM_EPINT = 1e-7,CPX_PARAM_TILIM = TL));
    @variable(mp,t0[i in pData.II] >= 0);
    @variable(mp,0 <= x0[i in pData.II, j in pData.Ji[i]] <= 1);
    @variable(mp,t[i in pData.II, ω in Ω] >= 0);
    @variable(mp,0 <= x[i in pData.II, j in pData.Ji[i], ω in Ω] <= 1);
    @variable(mp,G[i in pData.II, ω in Ω], Bin);
    @variable(mp,0 <= s[i in pData.II, j in pData.Ji[i], ω in Ω] <= 1);

    @constraint(mp, FConstr[i in pData.II, ω in Ω], disData[ω].H - (1-G[i,ω])*M[ω] <= t0[i]);
    @constraint(mp, GConstr[i in pData.II, ω in Ω], disData[ω].H - 1e-6 + G[i,ω]*M[ω] >= t0[i]);
    @constraint(mp, tConstr1[i in pData.II, ω in Ω], t[i,ω] + G[i,ω]*M[ω] >= t0[i]);
    @constraint(mp, tConstr2[i in pData.II, ω in Ω], t[i,ω] - G[i,ω]*M[ω] <= t0[i]);
    @constraint(mp, tConstr3[i in pData.II, ω in Ω], t[i,ω] >= disData[ω].H * G[i,ω]);
    @constraint(mp, xConstr1[i in pData.II, j in pData.Ji[i], ω in Ω], x[i,j,ω] + G[i,ω] >= x0[i,j]);
    @constraint(mp, xConstr2[i in pData.II, j in pData.Ji[i], ω in Ω], x[i,j,ω] - G[i,ω] <= x0[i,j]);
    @constraint(mp, durationConstr1[k in pData.K, ω in Ω], t[k[2],ω] - t[k[1],ω] >= pData.D[k[1]] + disData[ω].d[k[1]]*G[k[1],ω]
                      - sum(pData.D[k[1]]*pData.eff[k[1]][j]*x[k[1],j,ω] + disData[ω].d[k[1]]*pData.eff[k[1]][j]*s[k[1],j,ω] for j in pData.Ji[k[1]]));
    @constraint(mp, durationConstr2[k in pData.K], t0[k[2]] - t0[k[1]] >= pData.D[k[1]]*(1 - sum(pData.eff[k[1]][j]*x0[k[1],j] for j in pData.Ji[k[1]])));
    @constraint(mp, xConstr[i in pData.II,ω in Ω], sum(x[i,j,ω] for j in pData.Ji[i]) <= 1);
    @constraint(mp, GGcons1[i in pData.II, ω in 1:(length(Ω) - 1)], G[i,ω] >= G[i,ω + 1]);
    @constraint(mp, GGcons2[i in pData.K, ω in Ω], G[k[2],ω] >= G[k[1],ω]);
    @constraint(mp, budgetConstr[ω in Ω], sum(sum(pData.b[i][j]*x[i,j,ω] for j in pData.Ji[i]) for i in pData.II) <= pData.B);
    @constraint(mp, xConstr0[i in pData.II], sum(x0[i,j] for j in pData.Ji[i]) <= 1);
    @constraint(mp, budgetConstr0, sum(sum(pData.b[i][j]*x0[i,j] for j in pData.Ji[i]) for i in pData.II) <= pData.B);
    @constraint(mp, Slinear1[i in pData.II, j in pData.Ji[i], ω in Ω], s[i,j,ω] <= G[i,ω]);
    @constraint(mp, Slinear2[i in pData.II, j in pData.Ji[i], ω in Ω], s[i,j,ω] <= x[i,j,ω]);
    @constraint(mp, Slinear3[i in pData.II, j in pData.Ji[i], ω in Ω], s[i,j,ω] >= x[i,j,ω] - 1 + G[i,ω]);

    @objective(mp,Min,pData.p0*t0[0] + sum(disData[ω].prDis*t[0,ω] for ω in Ω));

    solve(mp);
    text = getvalue(mp[:t0]);
    xext = getvalue(mp[:x0]);
    gext = getvalue(mp[:G]);
    fext = getobjectivevalue(mp);
    return text,xext,fext,gext,mp;
end

function extForm_cheat(pData,disData,Ω,prec = 1e-4,TL = Inf)
    M = Dict();
    for ω in Ω
        M[ω] = sum(max(pData.D[i],pData.D[i]+disData[ω].d[i]) for i in pData.II if i != 0);
    end

    mp = Model(solver = GurobiSolver(IntFeasTol = 1e-7, TimeLimit = TL,MIPGap = prec));
    # mp = Model(solver = CplexSolver(CPX_PARAM_EPRHS = 1e-7,CPX_PARAM_EPINT = 1e-7,CPX_PARAM_TILIM = TL));
    @variable(mp,t0[i in pData.II] >= 0);
    @variable(mp,0 <= x0[i in pData.II, j in pData.Ji[i]] <= 1);
    @variable(mp,t[i in pData.II, ω in Ω] >= 0);
    @variable(mp,0 <= x[i in pData.II, j in pData.Ji[i], ω in Ω] <= 1);
    @variable(mp,G[i in pData.II, ω in Ω], Bin);
    @variable(mp,0 <= s[i in pData.II, j in pData.Ji[i], ω in Ω] <= 1);

    @constraint(mp, FConstr[i in pData.II, ω in Ω], disData[ω].H - (1 - G[i,ω])*M[ω] <= t0[i]);
    @constraint(mp, GConstr[i in pData.II, ω in Ω], disData[ω].H + G[i,ω]*M[ω] >= t0[i]);
    @constraint(mp, tConstr1[i in pData.II, ω in Ω], t[i,ω] + G[i,ω]*M[ω] >= t0[i]);
    @constraint(mp, tConstr2[i in pData.II, ω in Ω], t[i,ω] - G[i,ω]*M[ω] <= t0[i]);
    @constraint(mp, tConstr3[i in pData.II, ω in Ω], t[i,ω] >= disData[ω].H * G[i,ω]);
    @constraint(mp, xConstr1[i in pData.II, j in pData.Ji[i], ω in Ω], x[i,j,ω] + G[i,ω] >= x0[i,j]);
    @constraint(mp, xConstr2[i in pData.II, j in pData.Ji[i], ω in Ω], x[i,j,ω] - G[i,ω] <= x0[i,j]);
    @constraint(mp, durationConstr1[k in pData.K, ω in Ω], t[k[2],ω] - t[k[1],ω] >= pData.D[k[1]] + disData[ω].d[k[1]]*G[k[1],ω]
                      - sum(pData.D[k[1]]*pData.eff[k[1]][j]*x[k[1],j,ω] + disData[ω].d[k[1]]*pData.eff[k[1]][j]*s[k[1],j,ω] for j in pData.Ji[k[1]]));
    @constraint(mp, durationConstr2[k in pData.K], t0[k[2]] - t0[k[1]] >= pData.D[k[1]]*(1 - sum(pData.eff[k[1]][j]*x0[k[1],j] for j in pData.Ji[k[1]])));
    @constraint(mp, xConstr[i in pData.II,ω in Ω], sum(x[i,j,ω] for j in pData.Ji[i]) <= 1);
    @constraint(mp, GGcons1[i in pData.II, ω in 1:(length(Ω) - 1)], G[i,ω] >= G[i,ω + 1]);
    @constraint(mp, GGcons2[k in pData.K, ω in Ω], G[k[2],ω] >= G[k[1],ω]);
    @constraint(mp, budgetConstr[ω in Ω], sum(sum(pData.b[i][j]*x[i,j,ω] for j in pData.Ji[i]) for i in pData.II) <= pData.B);
    @constraint(mp, xConstr0[i in pData.II], sum(x0[i,j] for j in pData.Ji[i]) <= 1);
    @constraint(mp, budgetConstr0, sum(sum(pData.b[i][j]*x0[i,j] for j in pData.Ji[i]) for i in pData.II) <= pData.B);
    @constraint(mp, Slinear1[i in pData.II, j in pData.Ji[i], ω in Ω], s[i,j,ω] <= G[i,ω]);
    @constraint(mp, Slinear2[i in pData.II, j in pData.Ji[i], ω in Ω], s[i,j,ω] <= x[i,j,ω]);
    @constraint(mp, Slinear3[i in pData.II, j in pData.Ji[i], ω in Ω], s[i,j,ω] >= x[i,j,ω] - 1 + G[i,ω]);

    # for the same H, bind the G variables
    HDiff = Dict();
    for ω in Ω
        if disData[ω].H in keys(HDiff)
            push!(HDiff[disData[ω].H], ω);
        else
            HDiff[disData[ω].H] = [ω];
        end
    end
    for hIter in keys(HDiff)
        if length(HDiff[hIter]) > 1
            for item in 2:length(HDiff[hIter])
                for i in pData.II
                    @constraint(mp, G[i,HDiff[hIter][item]] == G[i,HDiff[hIter][1]]);
                end
            end
        end
    end

    @objective(mp,Min,pData.p0*t0[0] + sum(disData[ω].prDis*t[0,ω] for ω in Ω));

    solve(mp);

    text = Dict();
    xext = Dict();
    gext = Dict();
    for i in pData.II
        text[i] = getvalue(mp[:t0][i]);
        for j in pData.Ji[i]
            xext[i,j] = getvalue(mp[:x0][i,j]);
        end
        for ω in Ω
            gext[i,ω] = getvalue(mp[:G][i,ω]);
        end
    end
    fext = getobjectivevalue(mp);
    return text,xext,fext,gext,mp;
end

File: src/archive/python/extForm.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Jan 21 17:52:06 2019

@author: haoxiangyang
"""
from gurobipy import *

# deterministic formulations

def detBuild(pData):
    # construct the deterministic crashing problem
    mp = Model('deterministic_model')
    mp.setParam('OutputFlag',0)
    
    # set up variables
    t = mp.addVars(pData.II, lb = 0, name = "t")
    JJ = [(i,j) for i in pData.II for j in pData.Ji[i]]
    x = mp.addVars(JJ, lb = 0, ub = 1, name = "x")
    
    # set up constraints
    durationConstr = mp.addConstrs((t[k[1]] - t[k[0]] >= pData.D[k[0]]*(1 - sum(pData.eff[k[0]][j]*x[k[0],j] for j in pData.Ji[k[0]])) for k in pData.K), name = "durationConstr")
    budgetConstr = mp.addConstr((sum(sum(pData.b[i][j]*x[i,j] for j in pData.Ji[i]) for i in pData.II) <= pData.B), name = "budgetConstr")
    xConstr = mp.addConstrs((sum(x[i,j] for j in pData.Ji[i]) <= 1 for i in pData.II), name = "xConstr")
    mp.setObjective(t[0],GRB.MINIMIZE)
    mp.update()
    mp.optimize()
    
    tdet = {}
    xdet = {}
    fdet = mp.objVal
    for i in pData.II:
        tdet[i] = t[i].X
    for j in JJ:
        xdet[j] = x[j].X
    return tdet,xdet,fdet

def expModel(pData,eH,ed,M = 9999999):
    # script for two-stage expected time expected magnitude disruption program
    mp = Model('expected_model')
    mp.Params.OutputFlag = 0
    
    # set up variables
    t0 = mp.addVars(pData.II, lb = 0, name = "t0")
    JJ = [(i,j) for i in pData.II for j in pData.Ji[i]]
    x0 = mp.addVars(JJ, lb = 0, ub = 1, name = "x0")
    t = mp.addVars(pData.II, lb = 0, name = "t")
    x = mp.addVars(JJ, lb = 0, ub = 1, name = "x")
    G = mp.addVars(pData.II, vtype=GRB.BINARY, name = "G")
    s = mp.addVars(JJ, lb = 0, ub = 1, name = "s")

    # set up constraints
    durationConstr0 = mp.addConstrs((t0[k[1]] - t0[k[0]] >= pData.D[k[0]]*(1 - sum(pData.eff[k[0]][j]*x0[k[0],j] for j in pData.Ji[k[0]])) for k in pData.K), name = "durationConstr0")
    budgetConstr0 = mp.addConstr((sum(sum(pData.b[i][j]*x0[i,j] for j in pData.Ji[i]) for i in pData.II) <= pData.B), name = "budgetConstr0")
    xConstr0 = mp.addConstrs((sum(x0[i,j] for j in pData.Ji[i]) <= 1 for i in pData.II), name = "xConstr0")
    
    durationConstr = mp.addConstrs((t[k[1]] - t[k[0]] >= pData.D[k[0]] + ed[k[0]]*G[k[0]] - \
                               sum(pData.D[k[0]]*pData.eff[k[0]][j]*x[k[0],j] + ed[k[0]]*pData.eff[k[0]][j]*s[k[0],j] for j in pData.Ji[k[0]]) \
                               for k in pData.K), name = "durationConstr")
    budgetConstr = mp.addConstr((sum(sum(pData.b[i][j]*x[i,j] for j in pData.Ji[i]) for i in pData.II) <= pData.B), name = "budgetConstr")
    xConstr = mp.addConstrs((sum(x[i,j] for j in pData.Ji[i]) <= 1 for i in pData.II), name = "xConstr")
    
    FConstr = mp.addConstrs((eH - (1 - G[i])*M <= t0[i] for i in pData.II), name = "FConstr")
    GConstr = mp.addConstrs((eH + G[i]*M >= t0[i] for i in pData.II), name = "GConstr")
    
    tConstr1 = mp.addConstrs((t[i] + G[i]*M >= t0[i] for i in pData.II), name = "tConstr1")
    tConstr2 = mp.addConstrs((t[i] - G[i]*M <= t0[i] for i in pData.II), name = "tConstr2")
    tConstr3 = mp.addConstrs((t[i] >= G[i]*eH for i in pData.II), name = "tConstr3")
    
    xConstr1 = mp.addConstrs((x[i,j] + G[i] >= x0[i,j] for i in pData.II for j in pData.Ji[i]), name = "xConstr1")
    xConstr2 = mp.addConstrs((x[i,j] - G[i] <= x0[i,j] for i in pData.II for j in pData.Ji[i]), name = "xConstr2")

    sLinear1 = mp.addConstrs((s[i,j] <= G[i] for i in pData.II for j in pData.Ji[i]), name = "sLinear1")
    sLinear2 = mp.addConstrs((s[i,j] <= x[i,j] for i in pData.II for j in pData.Ji[i]), name = "sLinear2")
    sLinear3 = mp.addConstrs((s[i,j] >= G[i] + x[i,j] - 1 for i in pData.II for j in pData.Ji[i]), name = "sLinear3")
    
    mp.setObjective(pData.p0*t0[0] + (1 - pData.p0)*t[0],GRB.MINIMIZE)
    mp.update()
    mp.optimize()
    
    texp = {}
    xexp = {}
    gexp = {}
    fexp = mp.objVal
    for i in pData.II:
        texp[i] = t0[i].X
        for j in pData.Ji[i]:
            xexp[i,j] = x0[i,j].X
        gexp[i] = G[i].X
    return texp,xexp,fexp,gexp,mp

    
def extForm(pData,disData,Omega,prec = 1e-4,TL = 36000):
    # extensive formulation, time limit 10 hours
    M = {}
    for omega in Omega:
        M[omega] = sum(max(pData.D[i],pData.D[i]+disData[omega].d[i]) for i in pData.II if i != 0)
        
    mp = Model('extensive_model')
    mp.Params.IntFeasTol = 1e-8
    mp.Params.TimeLimit = TL
    mp.Params.MIPGap = prec
    mp.Params.OutputFlag = 0
    
    HDiff = {}
    for omega in Omega:
        if disData[omega].H in HDiff.keys():
            HDiff[disData[omega].H].append(omega)
        else:
            HDiff[disData[omega].H] = [omega]
    HDict = {}
    HList = []
    for hIter in HDiff.keys():
        for item in range(len(HDiff[hIter])):
            HDict[HDiff[hIter][item]] = HDiff[hIter][0]
        HList.append(HDiff[hIter][0])
    HList = sorted(HList)
    
    t0 = mp.addVars(pData.II, lb = 0, name = "t0")
    JJ = [(i,j) for i in pData.II for j in pData.Ji[i]]
    x0 = mp.addVars(JJ, lb = 0, ub = 1, name = "x0")
    t = mp.addVars(pData.II, Omega, lb = 0, name = "t")
    x = mp.addVars(JJ, Omega, lb = 0, ub = 1, name = "x")
    G = mp.addVars(pData.II, HList, vtype=GRB.BINARY, name = "G")
    s = mp.addVars(JJ, Omega, lb = 0, ub = 1, name = "s")
    
    durationConstr0 = mp.addConstrs((t0[k[1]] - t0[k[0]] >= pData.D[k[0]]*(1 - sum(pData.eff[k[0]][j]*x0[k[0],j] for j in pData.Ji[k[0]])) for k in pData.K), name = "durationConstr0")
    budgetConstr0 = mp.addConstr((sum(sum(pData.b[i][j]*x0[i,j] for j in pData.Ji[i]) for i in pData.II) <= pData.B), name = "budgetConstr0")
    xConstr0 = mp.addConstrs((sum(x0[i,j] for j in pData.Ji[i]) <= 1 for i in pData.II), name = "xConstr0")
    
    durationConstr = mp.addConstrs((t[k[1],omega] - t[k[0],omega] >= pData.D[k[0]] + disData[omega].d[k[0]]*G[k[0],HDict[omega]] - \
                                   sum(pData.D[k[0]]*pData.eff[k[0]][j]*x[k[0],j,omega] + disData[omega].d[k[0]]*pData.eff[k[0]][j]*s[k[0],j,omega] for j in pData.Ji[k[0]]) \
                                   for k in pData.K \
                                   for omega in Omega), name = "durationConstr")
    budgetConstr = mp.addConstrs((sum(sum(pData.b[i][j]*x[i,j,omega] for j in pData.Ji[i]) for i in pData.II) <= pData.B for omega in Omega), name = "budgetConstr")
    xConstr = mp.addConstrs((sum(x[i,j,omega] for j in pData.Ji[i]) <= 1 for i in pData.II for omega in Omega), name = "xConstr")
    
    FConstr = mp.addConstrs((disData[omega].H - (1 - G[i,HDict[omega]])*M[omega] <= t0[i] for i in pData.II for omega in Omega), name = "FConstr")
    GConstr = mp.addConstrs((disData[omega].H + G[i,HDict[omega]]*M[omega] >= t0[i] for i in pData.II for omega in Omega), name = "GConstr")
    
    tConstr1 = mp.addConstrs((t[i,omega] + G[i,HDict[omega]]*M[omega] >= t0[i] for i in pData.II for omega in Omega), name = "tConstr1")
    tConstr2 = mp.addConstrs((t[i,omega] - G[i,HDict[omega]]*M[omega] <= t0[i] for i in pData.II for omega in Omega), name = "tConstr2")
    tConstr3 = mp.addConstrs((t[i,omega] >= G[i,HDict[omega]]*disData[omega].H for i in pData.II for omega in Omega), name = "tConstr3")
    
    xConstr1 = mp.addConstrs((x[i,j,omega] + G[i,HDict[omega]] >= x0[i,j] for i in pData.II for j in pData.Ji[i] for omega in Omega), name = "xConstr1")
    xConstr2 = mp.addConstrs((x[i,j,omega] - G[i,HDict[omega]] <= x0[i,j] for i in pData.II for j in pData.Ji[i] for omega in Omega), name = "xConstr2")
    
    GGConstr1 = mp.addConstrs((G[i,HList[omega]] >= G[i,HList[omega+1]] for i in pData.II for omega in range(len(HList) - 1)), name = "GGConstr1")
    GGConstr2 = mp.addConstrs((G[k[1],omega] >= G[k[0],omega] for k in pData.K for omega in HList), name = "GGConstr2")
    
    sLinear1 = mp.addConstrs((s[i,j,omega] <= G[i,HDict[omega]] for i in pData.II for j in pData.Ji[i] for omega in Omega), name = "sLinear1")
    sLinear2 = mp.addConstrs((s[i,j,omega] <= x[i,j,omega] for i in pData.II for j in pData.Ji[i] for omega in Omega), name = "sLinear2")
    sLinear3 = mp.addConstrs((s[i,j,omega] >= G[i,HDict[omega]] + x[i,j,omega] - 1 for i in pData.II for j in pData.Ji[i] for omega in Omega), name = "sLinear3")
    
    mp.setObjective(pData.p0*t0[0] + sum(disData[omega].prDis*t[0,omega] for omega in Omega),GRB.MINIMIZE)
    mp.update()
    mp.optimize()
    
    text = {}
    xext = {}
    gext = {}
    fext = mp.objVal
    for i in pData.II:
        text[i] = t0[i].X
        for j in pData.Ji[i]:
            xext[i,j] = x0[i,j].X
        for omega in HList:
            gext[i,omega] = G[i,omega].X
    return text,xext,fext,gext,mp
    